docs(api): Revisions to Positions Within the OT-2 (#13287)
OpenTrons/opentrons-api: api/docs/v2/robot_position.rst

.. _robot-position:

##########################
Labware and Deck Positions

> ##########################

Reviewer: asterisks for H1

> The API automatically determines how the robot needs to move when working with the instruments and labware in your protocol. But sometimes you need direct control over these activities. The API lets you do just that. Specifically, you can control movements relative to labware and deck locations. You can also manage the gantry’s speed and trajectory as it traverses the enclosure. This document explains how to use API commands to take direct control of the robot and position it exactly where you need it.

Reviewer: "working area" instead of "enclosure"?

Author: Yes! Great suggestion. Adding that.

.. _position-relative-labware:

Position Relative to Labware
============================

> When the robot positions itself relative to a piece of labware, the labware definition determines movement, distance, height, offsets, and the actions it can perform. This section describes how these positional components are calculated and how to change them.

Reviewer: This over-collapses these concepts. The labware definition determines only a portion of the behavior.
- LPC determines labware offsets
- "type of action the robot will perform" determines whether the tip goes to the top or bottom of the well — it differs for aspirate and dispense.

Author: Ah, rereading this, I understand now. Will revise, something akin to the original:

> When the robot positions itself relative to a piece of labware, where it moves is determined by the labware definition, the actions you want it to perform, and the labware offsets for a specific deck slot. This section describes how these positional components are calculated and how to change them.

Top, Bottom, and Center
=======================

> Every well on every piece of labware has three addressable positions: top, bottom, and center. The position is determined by the labware definition and whether the labware is on a module or in a deck slot. You can use these positions as-is or calculate other positions relative to them.

Reviewer: ```suggestion
Every well on every piece of labware has three addressable positions: top, bottom, and center. The position is determined by the labware definition and what the labware is loaded on top of. You can use these positions as-is or calculate other positions relative to them.
```
Something like this? Generalizes now that we have the module/adapter split and `load_adapter()`.

Top
---

> As an example, let's look at the :py:meth:`.Well.top` method. It returns a position level with the top of the well, centered in both horizontal directions.

Reviewer: I don't think the "As an example" text does much besides making this longer to read.

.. versionadded:: 2.0

Bottom
------

As an example, lets look at the :py:meth:`.Well.bottom` method. It returns a position level with the bottom of the well, centered in both horizontal directions. 

.. code-block:: python

   plate['A1'].bottom()  # the bottom center of the well

> This is a good position to for :ref:`aspirating liquid <new-aspirate>` or an activity where you want the tip to contact the liquid. Similar to the ``Well.top()`` method, you can adjust the height of this position with the optional argument ``z``, which is measured in mm. Positive ``z`` numbers move the position up, negative ``z``` numbers move it down.

Reviewer: ```suggestion
This is a good position for :ref:`aspirating liquid <new-aspirate>` or an activity where you want the tip to contact the liquid. Similar to the ``Well.top()`` method, you can adjust the height of this position with the optional argument ``z``, which is measured in mm. Positive ``z`` numbers move the position up, negative ``z``` numbers move it down.
```

    # aspirate 1 mm above the bottom of the well (default)
    pipette.aspirate(50, plate['A1'])
    # dispense 1 mm above the bottom of the well (default)
    pipette.dispense(50, plate['A1'])

    # change clearance for aspiration to 2 mm
    pipette.well_bottom_clearance.aspirate = 2
    # aspirate 2 mm above the bottom of the well
    pipette.aspirate(50, plate['A1'])
    # still dispensing 1 mm above the bottom
    pipette.dispense(50, plate['A1'])

    pipette.aspirate(50, plate['A1'])
    # change clearance for dispensing to 10 mm      
    pipette.well_bottom_clearance.dispense = 10
    # dispense high above the well
    pipette.dispense(50, plate['A1'])

.. versionadded:: 2.0

.. _using_lpc:

Using Labware Position Check
============================

All positions relative to labware are adjusted automatically based on labware offset data. Calculate labware offsets by running Labware Position Check during protocol setup, either in the Opentrons App or on the Flex touchscreen. Version 6.0.0 and later of the robot software can apply previously calculated offsets on the same robot for the same labware type and deck slot, even across different protocols.

> You should only adjust labware offsets in your Python code if you plan to run your protocol in Jupyter notebook or with ``opentrons_execute``. This is because you can't run Labware Position Check from Jupyter or the command line. For these applications, calculate and apply labware offsets by:

Reviewer: @jwwojak Here is where to link to Advanced Control (replacing the remainder of this section). Then take the text from my gist and put it in `new_advanced_running.rst` between the Running a Previously Written Protocol and Using Custom Labware sections.

Modifying these attributes will affect all subsequent aspirate and dispense actions performed by the attached pipette, even those executed as part of a :py:meth:`.transfer` operation. This snippet from a sample protocol demonstrates how to work with and change the default clearance::

    # aspirate 1 mm above the bottom of the well (default)
    pipette.aspirate(50, plate['A1'])
    # dispense 1 mm above the bottom of the well (default)
    pipette.dispense(50, plate['A1'])

    # change clearance for aspiration to 2 mm
    pipette.well_bottom_clearance.aspirate = 2
    # aspirate 2 mm above the bottom of the well
    pipette.aspirate(50, plate['A1'])
    # still dispensing 1 mm above the bottom
    pipette.dispense(50, plate['A1'])

    pipette.aspirate(50, plate['A1'])
    # change clearance for dispensing to 10 mm      
    pipette.well_bottom_clearance.dispense = 10
    # dispense high above the well
    pipette.dispense(50, plate['A1'])

.. versionadded:: 2.0

.. _using_lpc:

Using Labware Position Check
============================

> All positions relative to labware are adjusted automatically based on the labware's x, y, z axis offset. The best way to calculate and apply these offsets is by using Labware Position Check when you run your protocol in the Opentrons App. With v6.0 of the app, you can apply previously calculated offsets on the same robot for the same labware type and deck slot, even with different protocols.

Reviewer: Let's make this cover Flex and the touchscreen as well.
```suggestion
All positions relative to labware are adjusted automatically based on labware offset data. Calculate labware offsets by running Labware Position Check during protocol setup, either in the Opentrons App or on the Flex touchscreen. Version 6.0.0 and later of the robot software can apply previously calculated offsets on the same robot for the same labware type and deck slot, even across different protocols.
```

> You shouldn't adjust labware offsets in your Python code if you plan to run your protocol in the app. However, if you are running your protocol in Jupyter notebook or with ``opentrons_execute``, the Labware Position Check is not directly available. For these applications, you can calculate and apply labware offsets by:

Reviewer: ```suggestion
You should only adjust labware offsets in your Python code if you plan to run your protocol in Jupyter notebook or with ``opentrons_execute``. This is because you can't run Labware Position Check from Jupyter or the command line. For these applications, calculate and apply labware offsets by:
```

touchscreen makes this confusing. could add something about it but then it gets wordy. cutting to the chase instead.

You shouldn't adjust labware offsets in your Python code if you plan to run your protocol in the app. However, if you are running your protocol in Jupyter notebook or with ``opentrons_execute``, the Labware Position Check is not directly available. For these applications, you can calculate and apply labware offsets by:
	
	1. Creating a "dummy" protocol that loads your labware and has each used pipette pick up a tip from a tip rack
	2. Importing the dummy protocol to the Opentrons App
	3. Running Labware Position Check
	4. Adding the offsets to your protocol
	1. Creating a "dummy" protocol that loads your labware and has each used pipette pick up a tip from a tip rack.
	2. Importing the dummy protocol to the Opentrons App.
	3. Running Labware Position Check.
	4. Adding the offsets to your protocol.
	
To prepare code written for Jupyter notebook so it can be run in the app, you need to include a metadata block and a ``run()`` function. And to enable Labware Position Check, you need to add a :py:meth:`.pick_up_tip` action for each pipette the protocol uses. For example, a dummy protocol using a P300 Single-Channel pipette, a reservoir, and a well plate would look like this:

> .. code-block:: python

Reviewer: We definitely need to keep this code block. This is the requisite "dummy protocol", without which you can't run LPC. 

I'm going to have a think about the broader purpose and structure of this section. I may put together an alternative in a gist or something.

> To prepare code written for Jupyter notebook and run it in the app, you need to specify an API level (see the :ref:`tutorial`) and put your commands in a ``run()`` function. To enable the Labware Position Check, you need to add a :py:meth:`.pick_up_tip` action for each pipette the protocol uses. For example, say you have a protocol that uses a Flex pipette, a tip rack, a 15 mL reservoir, and a 96-well plate. After importing this protocol to the Opentrons App, run Labware Position Check to get the x, y, and z offsets for the tip rack and labware. When complete, you can click **Get Labware Offset Data** to view automatically generated code that uses :py:meth:`.set_offset` to apply the offsets to each piece of labware.

Reviewer: could also link to "Specifying Versions". would require adding an anchor to versioning.rst

> To prepare code written for Jupyter notebook and run it in the app, you need to specify an API level (see the :ref:`tutorial`) and put your commands in a ``run()`` function. To enable the Labware Position Check, you need to add a :py:meth:`.pick_up_tip` action for each pipette the protocol uses. For example, say you have a protocol that uses a Flex pipette, a tip rack, a 15 mL reservoir, and a 96-well plate. After importing this protocol to the Opentrons App, run Labware Position Check to get the x, y, and z offsets for the tip rack and labware. When complete, you can click **Get Labware Offset Data** to view automatically generated code that uses :py:meth:`.set_offset` to apply the offsets to each piece of labware.

Reviewer: we basically never use the definite article "the" with "Labware Position Check" 📱

> The robot's base coordinate system is known as *deck coordinates*. Many API functions use this coordinate system, and you can also reference it directly. It is a right-handed coordinate system always specified in mm, with the origin ``(0, 0, 0)`` at the front left of the robot. The positive ``x`` direction is to the right, the positive ``y`` direction is to the back, and the positive ``z`` direction is up. 

Reviewer: I'm double-checking that this system is the same for Flex and OT-2.

Reviewer: Confirmed with Seth, they're the same at the level of PAPI.

    There are technically multiple vertical axes. For example, ``z`` is the axis of the left pipette mount and ``a`` is the axis of the right pipette mount. There are also pipette plunger axes: ``b`` (left) and ``c`` (right). You usually don't have to refer to these axes directly, since most motion commands are issued to a particular pipette and the robot automatically selects the correct axis to move. Similarly, :py:class:`.types.Location` only deals with ``x``, ``y``, and ``z`` values. 

Independent Movement
====================

For convenience, many methods have location arguments and incorporate movement automatically. This section will focus on moving the pipette independently, without performing other actions like ``aspirate()`` or ``dispense()``.

Move To
-------

> The :py:meth:`.InstrumentContext.move_to` method moves a pipette to any reachable location on the deck. If the pipette has picked up a tip, it will move the end of the tip to that position; if it hasn't, it will move the pipette nozzle to that position. As with all movement in a protocol, the robot calculates where to move in physical space by using its `pipette offset and tip length calibration <https://support.opentrons.com/s/article/Get-started-Calibrate-tip-length-and-pipette-offset>`_ data.

Reviewer: Let's strike the last sentence here. It's different for Flex and I don't think this level of detail is particularly illuminating. The prior sentence explains the behavior completely.

.. versionadded:: 2.0

Movement Speeds
===============

In addition to instructing the robot where to move a pipette, you can also control the speed at which it moves. Speed controls can be applied either to all pipette motions or to movement along a particular axis.

.. _gantry_speed: 

Gantry Speed
------------

The robot's gantry usually moves as fast as it can given its construction. The default speed for Flex varies between 300 and 350 mm/s. The OT-2 default is 400 mm/s. However, some experiments or liquids may require slower movements. In this case, you can reduce the gantry speed for a specific pipette by setting :py:obj:`.InstrumentContext.default_speed` like this::
        
	
	pipette.move_to(plate['A1'].top())  # move to the first well at default speed
	pipette.default_speed = 100         # reduce pipette speed
	pipette.move_to(plate['D6'].top())  # move to the last well at the slower speed

.. warning::

> 	These default speeds were chosen because they're the maximum speed Opentrons knows will work with the gantry. Your robot may be able to move faster, but you shouldn't increase this value unless instructed by Opentrons Support.

Reviewer: ```suggestion
	These default speeds were chosen because they're the maximum speeds that Opentrons knows will work with the gantry. Your robot may be able to move faster, but you shouldn't increase this value unless instructed by Opentrons Support.
```
agreement + overt clause boundary

.. _axis_speed_limits:

Axis Speed Limits
-----------------

In addition to controlling the overall gantry speed, you can set speed limits for each of the individual axes: ``x`` (gantry left/right motion), ``y`` (gantry forward/back motion), ``z`` (left pipette up/down motion), and ``a`` (right pipette up/down motion). Unlike ``default_speed``, which is a pipette property, axis speed limits are stored in a protocol property :py:obj:`.ProtocolContext.max_speeds`; therefore the ``x`` and ``y`` values affect all movements by both pipettes. This property works like a dictionary, where the keys are axes, assigning a value to a key sets a max speed, and deleting a key or setting it to ``None`` resets that axis's limit to the default:

.. code-block:: python
    :substitutions:

> 	protocol.max_speeds['x'] = 50    # limit x-axis to 50 mm/s

Reviewer: yes, this spacing is better!
